### ISSUE TITLE
clippy::needless_update with Default..default() is not usable

### ISSUE THREAD

**Issue by rbtcollins** [rust-lang/rust-clippy#12997]
### Summary

this is a reprisal of #6323.

In #6323 the recommendation is to use #[non_exhaustive], but that leads to 
`error[E0639]: cannot create non-exhaustive struct using struct expression`

Leaving users with:
a) toil when config structs have fields added
b) having to write and maintain a Builder (or derive with e.g.  https://docs.rs/derive_builder/latest/derive_builder/ ) because the native ergonomic code is behind a lint
c) disable the lint and not have places where it would actually apply be linted

So my argument here is that if people will bring in a builder, this lint isn't actually lifting the bar or adding safety vs permitting the idiom with Default::default() specifically.

### Lint Name

clippy::needless_update

### Reproducer

I tried this code:

```rust
#[derive(Default)]
struct S {
    x: i32,
}
fn foo() -> S {
    S {
        x: 123,
        ..S::default()
    }
}
```

I saw this happen:

```
warning: struct update has no effect, all the fields in the struct have already been specified
```

I expected to see this happen:

```
```

### Version

```text
rustc 1.77.0 (aedd173a2 2024-03-17)
binary: rustc
commit-hash: aedd173a2c086e558c2b66d3743b344f977621a7
commit-date: 2024-03-17
host: x86_64-unknown-linux-gnu
release: 1.77.0
LLVM version: 17.0.6
```

**Comment by Alexendoo** [comment]
I agree with https://github.com/rust-lang/rust-clippy/issues/6323#issuecomment-739500601, I'd want this to be linted

The case where it's intentional seem uncommon enough that `#[allow]` is reasonable
